### PR TITLE
Model-viewset level permissions

### PR DESCRIPTION
**Pull request by why-not-try-calmer**

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -46,14 +46,14 @@ jobs:
...
-        docker compose run django python manage.py migrate --no-input
+        docker compose exec django python manage.py collectstatic --no-input
+        docker compose exec django python manage.py migrate --no-input
 
         # sprinkle some test data (refreshing computed fields is done from the command handler)
-        docker compose run django python manage.py init --data --users
+        docker compose exec django python manage.py init --data --users
 
-    - name: Healthcheck
-      run: wget --no-check-certificate https://localhost/oapif/collections/signalo_core.pole/items
```
Comment (olivierdalang) [inline .github/workflows/tests.yml:56]: If you remove this test, ensure that we still somehow cover this case (integration, incl. caddy).

Otherwise keep it, and run the django tests afterwards.

File: src/signalo/core/tests/test_model_permissions.py
Comment (olivierdalang) [inline src/signalo/core/tests/test_model_permissions.py:1]: You should start from https://www.django-rest-framework.org/api-guide/testing/#example

This includes DRF's tests classes (that extend django tests), which include stuff like a test client allowing to simulate requests, isolated edits to the database between tests, etc.

File: src/signalo/settings.py [hunk 2]
```diff
@@ -146,6 +146,7 @@
     # "geocity.auth.InternalTokenAuthentication",
     # ),
     "DEFAULT_PAGINATION_CLASS": "django_oapif.pagination.OapifPagination",
+    "DEFAULT_PERMISSION_CLASSES": ("rest_framework.permissions.AllowAny",),
```
Comment (olivierdalang) [inline src/signalo/settings.py:149]: This is super-permissive, I guess it's extremely rare that one wants just anyone to be able to edit their data (basically a wiki). As discussed, let's default to `[DjangoModelPermissionsOrAnonReadOnly](https://www.django-rest-framework.org/api-guide/permissions/#djangomodelpermissionsoranonreadonly)`.

Then you can add a new "WikiModel" class with `AllowAny` permission, and "SecretModel" with `IsAdminUse`, just to demonstrate that it works (add these to the `edge_cases` app)

File: src/signalo/core/tests/test_model_permissions.py
Comment (olivierdalang) [reply src/signalo/core/tests/test_model_permissions.py:1]: And about code organisation, since we organise the code in django apps already, which ideally don't grow too big, you could just keep a single `tests.py` file. In theory it shouldn't get too big, and we can also split later if it does.

File: README.md [hunk 3]
```diff
@@ -22,6 +22,18 @@ docker compose run django python manage.py init
...
+
+By default the viewsets under `signalo/core` use the `AllowAny` permissions class. You can override these app-level permissions for particular viewsets when registering their corresponding models. (Refer to https://www.django-rest-framework.org/api-guide/permissions/#api-reference for permission classes). Example:
+
+    models.py
+    ---------
+    @register_oapif_viewset(
+        custom_viewset_attrs={
+            "model_to_viewset_permissions": (permissions.DjangoModelPermissionsOrAnonReadOnly,)
+        }
+    )
```
Comment (olivierdalang) [inline README.md:35]: The intended usage is:
```
@register_oapif_viewset(
    custom_viewset_attrs={
        "permission_classes": (permissions.DjangoModelPermissionsOrAnonReadOnly,)
    }
)
```

(the keys are directly the DRF class attributes of the viewset, so we need less specific logic (see comment below in the decorator)

Also in the snippet, it helps to include the imports and a (partial) model definition (at least with the geom field), otherwise not very clear to what the decorator applies.

**Review by olivierdalang — CHANGES_REQUESTED** [review]

File: src/signalo/settings.py (hunk 2 above)
Comment (why-not-try-calmer) [reply src/signalo/settings.py:149]: > As with DjangoModelPermissions, this permission must only be applied to views that have a .queryset property or .get_queryset() method.

This condition is not met as of this PR.

Comment (olivierdalang) [reply src/signalo/settings.py:149]: Indeed didn't think about the RootView and CollectionView (that's what you're refering to, right ?)

Still, defaulting to wiki isn't good enough. Either we could 1/ still put the DjangoModelPermissionsOrAnonReadOnly as a default, and override with AllowAny on RootView and CollectionView. Otherwise we could 2/ introduce a new setting, called `OGCAPIF_DEFAULT_PERMISSION_CLASSES` (outside of the native DRF settings dict) that we would use as a default in our decorator or even better directly on the `OAPIFDescribeModelViewSetMixin`.

IMO better to go with 1/, it's more standard, less stuff to document. The only drawback is that it would be difficult to protect RootView and CollectionView in case one doesn't want to expose them to unlogged users (for RootView, I think we always want to expose it, as it may also include login endpoint, and for collectionview see my commend below)

**Comment by olivierdalang** [comment]
Another aspect about this, is that we also need to filter the collection view, as we don't want to list models there which the user cannot see (otherwise, he'll see them in the catalog, try to add them to his project, and get 403 errors, which isn't very useful).

IMO this should be part of this PR too. Can you give it a shot ?

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -52,6 +52,10 @@ jobs:
         # sprinkle some test data (refreshing computed fields is done from the command handler)
         docker compose exec django python manage.py init --data --users
 
+        # test 'signalo'
+        docker compose exec django python manage.py test signalo.core
+        docker compose exec django python manage.py test signalo.edge_cases
```
Comment (olivierdalang) [inline .github/workflows/test.yml:57]: not sure, but probably `docker compose exec django python manage.py test` or `docker compose exec django python manage.py test signalo` would be enough

File: README.md [hunk 5]
```diff
@@ -22,6 +22,20 @@ docker compose exec django python manage.py init
...
+
+    models.py
+    ---------
+    from rest_framework.permissions import DjangoModelPermissionsOrAnonReadOnly
+    ...
+    @register_oapif_viewset(
+        custom_viewset_attrs={
+            "permissions_classes": (DjangoModelPermissionsOrAnonReadOnly,)
+        }
+    )
```
Comment (olivierdalang) [inline README.md:37]: as said, please include a Model definition here, otherwise the decorator is hanging and it's not clear on what it applies

File: README.md [hunk 6]
```diff
@@ -22,6 +22,20 @@ docker compose exec django python manage.py init
 # Wait a little, then check that https://localhost/oapif/collections/signalo_core.pole/items works from your browser
 ```
 
+## Authentication & permissions
+
+By default the viewsets under `signalo/core` use the `AllowAny` permissions class. You can override these app-level permissions for particular viewsets when registering their corresponding models. (Refer to https://www.django-rest-framework.org/api-guide/permissions/#api-reference for permission classes). Example:
+
+    models.py
+    ---------
+    from rest_framework.permissions import DjangoModelPermissionsOrAnonReadOnly
```
Comment (olivierdalang) [inline README.md:31]: missing register_oapif_viewset import statement

File: README.md [hunk 7]
```diff
@@ -22,6 +22,20 @@ docker compose exec django python manage.py init
 # Wait a little, then check that https://localhost/oapif/collections/signalo_core.pole/items works from your browser
 ```
 
+## Authentication & permissions
+
+By default the viewsets under `signalo/core` use the `AllowAny` permissions class. You can override these app-level permissions for particular viewsets when registering their corresponding models. (Refer to https://www.django-rest-framework.org/api-guide/permissions/#api-reference for permission classes). Example:
```
Comment (olivierdalang) [inline README.md:27]: adapt to new logic

File: src/django_oapif/views.py [hunk 8]
```diff
@@ -80,7 +82,12 @@ def get(self, request, *args, **kwargs):
             # Instantiating the viewset
             viewset = Viewset(request=request)
 
-            collections.append(viewset._describe(request, base_url=f"collections/"))
+            # Returning only viewsets that match the request's permissions
+            if all(
+                permission.has_permission(request, self)
+                for permission in viewset.get_permissions()
+            ):
+                collections.append(viewset._describe(request, base_url=f"collections/"))
```
Comment (olivierdalang) [inline src/django_oapif/views.py:90]: nice !

File: src/signalo/core/models.py [hunk 9]
```diff
@@ -4,11 +4,16 @@
...
+from rest_framework import permissions
 
 from django_oapif.decorators import register_oapif_viewset
 
 
-@register_oapif_viewset()
+@register_oapif_viewset(
+    custom_viewset_attrs={
+        "permission_classes": (permissions.IsAuthenticatedOrReadOnly,)
+    }
```
Comment (olivierdalang) [inline src/signalo/core/models.py:15]: Leave to default, which better matches the use case we want to demonstrate.

File: src/signalo/core/models.py [hunk 10]
```diff
@@ -22,7 +27,11 @@ def _serialized(self):
         return f'{{"id": "{str(self.id)}", "type": "Feature", "geometry": {self.geom.geojson}, "properties": {{"name": "{self.name}"}}}}'
 
 
-@register_oapif_viewset()
+@register_oapif_viewset(
+    custom_viewset_attrs={
+        "permission_classes": (permissions.IsAuthenticatedOrReadOnly,)
+    }
```
Comment (olivierdalang) [inline src/signalo/core/models.py:33]: Same (Leave to default, which better matches the use case we want to demonstrate.)

File: src/signalo/edge_cases/models.py [hunk 11]
```diff
@@ -46,4 +47,20 @@ class DifferentSrid(models.Model):
...
 
     id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
-    geom = models.PointField(srid=2056)
+    geom = models.PointField(srid=2154)
+
+
+@register_oapif_viewset(
+    custom_serializer_attrs={"permission_classes": (permissions.AllowAny,)}
+)
+class AllowAnyModel(models.Model):
```
Comment (olivierdalang) [inline src/signalo/edge_cases/models.py:56]: Add a docstring to document what this is (that docstring is exposed as the layer's description)

File: src/signalo/edge_cases/models.py [hunk 12]
```diff
@@ -46,4 +47,20 @@ class DifferentSrid(models.Model):
...
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    geom = models.PointField(srid=2154)
+
+
+@register_oapif_viewset(
+    custom_serializer_attrs={"permission_classes": (permissions.IsAdminUser,)}
+)
+class IsAdminUserModel(models.Model):
```
Comment (olivierdalang) [inline src/signalo/edge_cases/models.py:64]: Same (Add a docstring to document what this is (that docstring is exposed as the layer's description))

File: src/signalo/edge_cases/tests.py [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+        response = self.client.get(
+            "/oapif/collections/signalo_edge_cases.allowanymodel"
+        )
+        self.assertIs(response.status_code, 200)
+
+    def test_isadminusermodel(self):
+        response = self.client.put(
+            "/oapif/collections/signalo_edge_cases.isadminusermodel", data={}
+        )
+        self.assertIn(response.status_code, [403, 405])
```
Comment (olivierdalang) [inline src/signalo/edge_cases/tests.py:15]: Here I think you could test all four CRUD operations, with three type of requests (non authenticated, authenticated without permissions, and authenticated with permissions), each time asserting either success and failure.

It's likely a bit redudent to DRF's tests, but since it's a very sensitive topic I think it's worht it

File: src/signalo/edge_cases/tests.py [hunk 14]
```diff
@@ -0,0 +1,15 @@
+from rest_framework.test import APITestCase
+
+
+class TestViewsets(APITestCase):
```
Comment (olivierdalang) [inline src/signalo/edge_cases/tests.py:4]: Can you add a test case (with the same users as described below) testing the "collections" view ?

File: src/signalo/edge_cases/models.py [hunk 15]
```diff
@@ -46,4 +47,20 @@ class DifferentSrid(models.Model):
...
+
+
+@register_oapif_viewset(
+    custom_serializer_attrs={"permission_classes": (permissions.AllowAny,)}
+)
+class AllowAnyModel(models.Model):
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    geom = models.PointField(srid=2154)
+
+
```
Comment (olivierdalang) [inline src/signalo/edge_cases/models.py:60]: After looking at the tests, I'd add another model with default permissions (`DefaultPermissionsModel`), so we can test DjangoModelPermissionsOrAnonReadOnly specifically

File: src/signalo/core/tests.py
Comment (olivierdalang) [inline src/signalo/core/tests.py:1]: We can drop all these tests, and rely on the tests of the edge cases which are more systematic.

**Review by olivierdalang — CHANGES_REQUESTED** [review]
Nice ! Some more comments

File: .github/workflows/test.yml (hunk 4 above)
Comment (olivierdalang) [reply .github/workflows/test.yml:57]: btw you can add this to the quickstart section of the readme too

**Comment by why-not-try-calmer** [comment]
Okay, thanks for the comments, will accomodate now

File: src/signalo/core/tests.py
Comment (why-not-try-calmer) [reply src/signalo/core/tests.py:1]: It raises a red flag in me to have two sister apps and one make the tests for the others. Why not regroup all tests under the same file / folder, or leave every app do its own laundry?

File: src/django_oapif/views.py [hunk 16]
```diff
@@ -80,7 +82,12 @@ def get(self, request, *args, **kwargs):
             # Instantiating the viewset
             viewset = Viewset(request=request)
 
-            collections.append(viewset._describe(request, base_url=f"collections/"))
+            # Returning only viewsets that match the request's permissions
+            if all(
+                permission.has_permission(request, self)
+                for permission in viewset.get_permissions()
+            ):
```
Comment (olivierdalang) [inline src/django_oapif/views.py:89]: Just thinking, are you sure it's supposed to be "all" and not "any" ? Doesn't matching one permission suffice in DRF ?

Actually, we shouldn't implement it ourselves, especially since this is quite a sensitive topic. I think you can use `viewset.check_permissions(request)` (defined on the APIView)  and catch `exceptions.NotAuthenticated` and `exceptions.PermissionDenied`

File: src/django_oapif/views.py [hunk 17]
```diff
@@ -83,11 +83,15 @@ def get(self, request, *args, **kwargs):
...
-            ):
+            try:
+                viewset.check_permissions(request)
                 collections.append(viewset._describe(request, base_url=f"collections/"))
+            except (
+                exceptions.NotAuthenticated,
+                exceptions.AuthenticationFailed,
+                exceptions.PermissionDenied,
+            ) as e:
+                return Response(status=e.status_code)
```
Comment (olivierdalang) [inline src/django_oapif/views.py:94]: That looks strange, shouldn't we just skip unpermitted items instead of returning a response ?

**Comment by why-not-try-calmer** [comment]
@olivierdalang Sorry to bother you; I'm getting a puzzling result:

```
 => Failed 2/16
list: /oapif/collections/signalo_edge_cases.testpermissionisadminusermodel/items (got 200, expected 403)
read: /oapif/collections/signalo_edge_cases.testpermissionisadminusermodel/items (got 200, expected 403)
```
The model under scrutiny uses the `IsAdmin` permission class, which I understand should be restricted even in read for the non-authenticated user, as per https://www.django-rest-framework.org/api-guide/permissions/#isadminuser. However, the results shown above suggest that the user is still able to `list` and `view` the items. Any idea why?

File: src/signalo/edge_cases/models.py [hunk 18]
```diff
@@ -46,4 +47,32 @@ class DifferentSrid(models.Model):
     """This model uses a different SRID from the others, allowing to check that this is properly managed by QGIS."""
 
     id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
-    geom = models.PointField(srid=2056)
+    geom = models.PointField(srid=2154)
+
+
+@register_oapif_viewset(
+    custom_serializer_attrs={"permission_classes": (permissions.AllowAny,)}
```
Comment (olivierdalang) [inline src/signalo/edge_cases/models.py:54]: custom_viewset_attrs not custom_serializer_attrs

File: src/signalo/edge_cases/management/commands/populate_edge_cases.py [hunk 19]
```diff
@@ -41,4 +46,11 @@ def handle(self, *args, **options):
         # Create data for DifferentSrid
         DifferentSrid.objects.create(geom=f"Point(2600000 1200000)")
 
+        # Create data for models whose permissions we are testing against
+        TestPermissionAllowAny.objects.create(geom=f"Point(2600000 1200000)")
+        TestPermissionDefaultPermissionsSettings.objects.create(
+            geom=f"Point(2600000 1200000)"
+        )
+        TestPermissionIsAdminUserModel.objects.create(geom=f"Point(2800000 1300000)")
```
Comment (why-not-try-calmer) [inline src/signalo/edge_cases/management/commands/populate_edge_cases.py:54]: For some reason I cannot identify this doesn't create any instance, making the tests proposed by this PR fail.

**Review by suricactus — APPROVED** [review]
Existing code looks good at first glance. Didn't go deep into the test matrix to fully understand the logic. While it would be good to avoid such test helpers, I believe there was a good reasoning behind it.

One remark - I saw a lot of dead code around, like "there will be follow-up PR". If the follow-up is not coming immediately, probably a more descriptive comments would help in the mid-term.

**Comment by why-not-try-calmer** [comment]
> Existing code looks good at first glance. Didn't go deep into the test matrix to fully understand the logic. While it would be good to avoid such test helpers, I believe there was a good reasoning behind it.
> 
> One remark - I saw a lot of dead code around, like "there will be follow-up PR". If the follow-up is not coming immediately, probably a more descriptive comments would help in the mid-term.

Thanks a lot for the review, good points, I'll clean up the commented code.
